include/revng/ABI/FunctionType.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Types.h"
+
+namespace abi::FunctionType {
+
+/// Best effort `CABIFunctionType` to `RawFunctionType` conversion.
+///
+/// If `ABI` is not specified, `TheBinaryToAddSupportTypesTo.DefaultABI`
+/// is used instead.
+std::optional<model::CABIFunctionType>
+tryConvertToCABI(const model::RawFunctionType &Function,
+                 model::Binary &TheBinaryToAddSupportTypesTo,
+                 std::optional<model::ABI::Values> ABI = std::nullopt);
```

Comment:
The fact that we return a `optional<CABIFunctionType>` but we need to get a non-`const` `model::Binary` is suspicious.
It seems to me we're risking to modify `Binary` but then the user might decide to discard the returned function type but it will still have changes on `model::Binary`.

I'd just return a `TypePath` or a `model::CABIFunctionType *` to a type that has been inserted.

Also:
```suggestion
std::optional<model::CABIFunctionType>
tryConvertToCABI(const model::RawFunctionType &Function,
                 model::Binary &TheBinaryToAddSupportTypesTo,
                 model::ABI::Values ABI = model::ABI::Invalid);
```

Reply:
1. Is a really good point I haven't thought about earlier, I feel like that can be fixed by temporarily storing "support types" internally before adding them to the binary at the very end. Should I apply the same approach to the primitive types, or is it fine if this registers a new primitive type even when failing?

2. Yeah, that's a good point and an easy fix, will do.

3. I guess I like the extra explicitness `std::optional` provides, this works too, though.

Reply:
1. In order not to increase complexity for this "small" problem, the ideal solution would be to first perform all the checks to verify whether we can return a result or not, and *then* modify `Binary`, once you're sure you'll be successful.
  For what concerns `PrimitiveType`s: feel free to add them, at one point we'll offer some guarantees on what `PrimitiveType`s you can assume will be already available. Put a TODO hinting at this and we'll fix this later.
2. :+1:
3. Yeah, extra explicit but now you have two distinct way to represent similar concepts. Either we drop `Invalid` everywhere in favor of `optional<myenum>` or we use `Invalid`.

Reply:
1. This would require a clear "condition" of what `CABI` function can't in fact be represented as a `Raw` function. Which is nice in it's own right, but this might cause a bit of a rewrite of the underlying implementation.

Reply:
1. We can have a function with a `bool` (template?) that when it's `false` "tries" and reports success/failure and when it's `true` actually does the thing? Not excited about this idea.